### PR TITLE
Arrow writing

### PR DESCRIPTION
**Pull request by Kopilov**
Saving DataFrame in Arrow format, as is or per schema.

Warnings are muted by default, this may be changed after #147 resolving.

### CODE REVIEW

File: dataframe-arrow/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowKtTest.kt [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+
+    @Test
+    fun testWritingGeneral() {
+        fun assertEstimation(citiesDeserialized: DataFrame<*>) {
+            citiesDeserialized["name"] shouldBe citiesExampleFrame["name"]
+            citiesDeserialized["affiliation"] shouldBe citiesExampleFrame["affiliation"]
+            citiesDeserialized["is_capital"] shouldBe citiesExampleFrame["is_capital"]
+            citiesDeserialized["population"] shouldBe citiesExampleFrame["population"]
+            citiesDeserialized["area"] shouldBe citiesExampleFrame["area"]
+            citiesDeserialized["settled"].type() shouldBe typeOf<LocalDate>() // cities["settled"].type() refers to FlexibleTypeImpl(LocalDate..LocalDate?) and does not match typeOf<LocalDate>()
```
Comment (koperagen) [inline dataframe-arrow/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowKtTest.kt:111]: Nice catch, i think it should be fixed in `readArrow` later as well

File: dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriter.kt [hunk 2]
```diff
@@ -0,0 +1,491 @@
...
+            check(actualField.isNullable)
+            infillWithNulls(vector, dataFrame.rowsCount())
+        } else {
+            infillVector(vector, convertedColumn)
+        }
+        return vector
+    }
+
+    private fun List<AnyCol>.toVectors(): List<FieldVector> = this.toArrowSchema(warningSubscriber).fields.mapIndexed { i, field ->
+        allocateVectorAndInfill(field, this[i], true, true)
```
Comment (koperagen) [inline dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriter.kt:296]: Maybe convert AnyCol to Field one by one, so you can get rid of this nullability in `this[i]`?

Comment (koperagen) [reply dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriter.kt:296]: like 
```
map { 
   val field = it.toArrowField() 
   allocateVectorAndInfill(field, it, true, true)
}
```

**Comment by koperagen** [comment]
Wow, hi! That's a lot of work, thank you for the PR. I'll leave a couple of comments about the details, but right now i seem to lack general understanding, so i'll review it again a bit later. 

What do you think about making possible warnings a sealed class with overloaded toString for default behaviour? So that you can access actual exception in conversion or DataColumn and KType, etc.

**Comment by koperagen** [comment]
Sorry, sorry :D I just kept this open while merging other PR. Can you re-open?

File: dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriter.kt (hunk 2 above)
Comment (Kopilov) [reply dataframe-arrow/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowWriter.kt:296]: Accepted, thanks

File: dataframe-arrow/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowKtTest.kt (hunk 1 above)
Comment (Kopilov) [reply dataframe-arrow/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/ArrowKtTest.kt:111]: `FlexibleTypeImpl` comes not from Arrow but from `listOf(LocalDate.of())` in the sample generation.
What exactly do you want to fix?

**Comment by Kopilov** [comment]
It seems that this can not be reopened. Duplicated in #169 by cherry-picking the branch manually.

**Pull request merged**